File: app/src/main/java/com/tunjid/rcswitchcontrol/model/RfSwitch.java
package com.tunjid.rcswitchcontrol.model;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static android.content.Context.MODE_PRIVATE;

/**
 * A model representing an RF switch
 * <p>
 * Created by tj.dahunsi on 3/11/17.
 */

public class RfSwitch implements Parcelable {

    // Shared preference key
    public static final String SWITCH_PREFS = "SwitchPrefs";
    private static final String SWITCHES_KEY = "Switches";

    private static final Gson gson = new Gson();

    private String name;

    private byte bitLength;
    private byte pulseLength;
    private byte protocol;

    private byte[] onCode = new byte[4];
    private byte[] offCode = new byte[4];

    private RfSwitch() {

    }

    public static ArrayList<RfSwitch> getSavedSwitches(Context context) {
        SharedPreferences sharedPreferences = context.getSharedPreferences(SWITCH_PREFS, MODE_PRIVATE);
        String jsonString = sharedPreferences.getString(SWITCHES_KEY, "");
        RfSwitch[] array = gson.fromJson(jsonString, RfSwitch[].class);

        return array == null ? new ArrayList<RfSwitch>() : new ArrayList<>(Arrays.asList(array));

    }

    public static void saveSwitches(Context context, List<RfSwitch> switches) {
        SharedPreferences sharedPreferences = context.getSharedPreferences(SWITCH_PREFS, MODE_PRIVATE);
        sharedPreferences.edit().putString(SWITCHES_KEY, gson.toJson(switches)).apply();
    }

    public String getName() {
        return name;
    }

    public byte[] getOnCode() {
        return onCode;
    }

    public byte[] getOffCode() {
        return offCode;
    }

    public byte getBitLength() {
        return bitLength;
    }

    public byte getPulseLength() {
        return pulseLength;
    }

    public byte getProtocol() {
        return protocol;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        RfSwitch rfSwitch = (RfSwitch) o;

        return Arrays.equals(onCode, rfSwitch.onCode) && Arrays.equals(offCode, rfSwitch.offCode);

    }

    @Override
    public int hashCode() {
        int result = Arrays.hashCode(onCode);
        result = 31 * result + Arrays.hashCode(offCode);
        return result;
    }

    private RfSwitch(Parcel in) {
        name = in.readString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name);
    }

    public static final Parcelable.Creator<RfSwitch> CREATOR = new Parcelable.Creator<RfSwitch>() {
        @Override
        public RfSwitch createFromParcel(Parcel in) {
            return new RfSwitch(in);
        }

        @Override
        public RfSwitch[] newArray(int size) {
            return new RfSwitch[size];
        }
    };

    public static final class SwitchCreator {
        State state;
        RfSwitch rfSwitch;

        public SwitchCreator() {
            state = State.ON_CODE;
        }

        public void withOnCode(byte[] code) {
            state = State.OFF_CODE;

            rfSwitch = new RfSwitch();
            rfSwitch.pulseLength = code[4];
            rfSwitch.bitLength = code[5];
            rfSwitch.protocol = code[6];

            System.arraycopy(code, 0, rfSwitch.onCode, 0, 4);
        }

        public RfSwitch withOffCode(byte[] code) {
            state = State.ON_CODE;

            int pulseLength = rfSwitch.pulseLength;
            pulseLength += (int) code[4];
            pulseLength /= 2;
            rfSwitch.pulseLength = (byte) pulseLength;

            System.arraycopy(code, 0, rfSwitch.offCode, 0, 4);
            return rfSwitch;
        }


        public State getState() {
            return state;
        }
    }

    public enum State {
        ON_CODE, OFF_CODE
    }
}

File: app/src/main/java/com/tunjid/rcswitchcontrol/data/RfSwitch.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.tunjid.rcswitchcontrol.data

import android.os.Parcel
import android.os.Parcelable
import android.util.Base64
import androidx.annotation.StringDef
import com.tunjid.rcswitchcontrol.nsd.protocols.SerialRFProtocol

/**
 * A model representing an RF switch
 *
 *
 * Created by tj.dahunsi on 3/11/17.
 */

class RfSwitch() : Parcelable, Device {

    override var name: String = "Switch"
    override val key: String
        get() = SerialRFProtocol::class.java.name

    private var bitLength: Byte = 0
    private var protocol: Byte = 0

    private var pulseLength = ByteArray(4)
    private var onCode = ByteArray(4)
    private var offCode = ByteArray(4)

    override val diffId
        get() = "$onCode-$offCode"

    @Retention(AnnotationRetention.SOURCE)
    @StringDef(ON_CODE, OFF_CODE)
    internal annotation class SwitchCode {}

    private constructor(`in`: Parcel) : this() {
        name = `in`.readString()!!
        protocol = `in`.readByte()
        bitLength = `in`.readByte()
        onCode = `in`.createByteArray()!!
        offCode = `in`.createByteArray()!!
        pulseLength = `in`.createByteArray()!!
    }

    private fun getTransmission(state: Boolean): ByteArray {
        val transmission = ByteArray(10)

        System.arraycopy(if (state) onCode else offCode, 0, transmission, 0, onCode.size)
        System.arraycopy(pulseLength, 0, transmission, 4, pulseLength.size)
        transmission[8] = bitLength
        transmission[9] = protocol

        return transmission
    }

    fun getEncodedTransmission(state: Boolean): String =
            Base64.encodeToString(getTransmission(state), Base64.DEFAULT)

    // Equals considers the code only, not the name
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val rcSwitch = other as RfSwitch? ?: return false

        return onCode.contentEquals(rcSwitch.onCode) && offCode.contentEquals(rcSwitch.offCode)
    }

    override fun hashCode(): Int {
        var result = onCode.contentHashCode()
        result = 31 * result + offCode.contentHashCode()
        return result
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeByte(protocol)
        dest.writeByte(bitLength)
        dest.writeByteArray(onCode)
        dest.writeByteArray(offCode)
        dest.writeByteArray(pulseLength)
    }

    class SwitchCreator {
        @SwitchCode
        @get:SwitchCode
        var state: String = ON_CODE
            internal set

        private lateinit var rfSwitch: RfSwitch

        fun withOnCode(code: ByteArray) {
            state = OFF_CODE

            rfSwitch = RfSwitch()

            rfSwitch.bitLength = code[8]
            rfSwitch.protocol = code[9]

            System.arraycopy(code, 0, rfSwitch.onCode, 0, 4)
            System.arraycopy(code, 4, rfSwitch.pulseLength, 0, 4)
        }

        fun withOffCode(code: ByteArray): RfSwitch {
            state = ON_CODE
            System.arraycopy(code, 0, rfSwitch.offCode, 0, 4)
            return rfSwitch
        }
    }

    companion object {
        // Shared preference key
        const val SWITCH_PREFS = "SwitchPrefs"

        const val ON_CODE = "on"
        const val OFF_CODE = "off"

        @JvmField
        @Suppress("unused")
        val CREATOR: Parcelable.Creator<RfSwitch> = object : Parcelable.Creator<RfSwitch> {
            override fun createFromParcel(`in`: Parcel): RfSwitch = RfSwitch(`in`)

            override fun newArray(size: Int): Array<RfSwitch?> = arrayOfNulls(size)
        }
    }
}
